source/core/src/main/com/csse3200/game/components/quests/AbstractQuest.java

```diff
@@ -0,0 +1,176 @@
+    /**
+     * True if quest is active (not completed or failed).
+     */
+    private boolean isActive;
+    /** Triggers for task completion. */
+    private final String[] taskCompletionTriggers;
+
+    /** Constructor design for implementing subclasses. */
+    protected AbstractQuest(String questName, String questDescription, List<Task> tasks, Boolean isAchievement,
+                         Boolean isSecretQuest, Map<DialogueKey,String[]> dialogue, String[] taskCompletionTriggers) {
+        this.questName = questName;
+        this.questDescription = questDescription;
+        this.tasks = tasks;
+        this.isAchievement = isAchievement;
+        this.isSecretQuest = isSecretQuest;
+        this.isActive = true;
+        this.questDialogue = dialogue;
+        this.taskCompletionTriggers = taskCompletionTriggers;
+    }
+
+    /** Returns quest name. */
+    public String getQuestName() {
+        return questName;
+    }
+    /** Returns task array for quest subtasks. */
+    public List<Task> getTasks() {
+        return tasks;
+    }
+    /** Returns true if quest is completed. */
+    public boolean isQuestCompleted() {
+        return currentTaskIndex >= tasks.size() && tasks.stream().allMatch(Task::isCompleted);
+    }
+    /** Returns number of quest subtasks completed. */
+    public int getProgression() {
+        return currentTaskIndex;
+    }
+
+    /**
+     * Returns the description of the current subtask being completed
+     * for a test or "QUEST COMPLETED" if the quest is completed.
+     */
+    public String getCurrentTaskDescription() {
+        if (isQuestCompleted()) {
+            return "QUEST COMPLETED";
+        }
+        if (currentTaskIndex == getNumQuestTasks()) {
+            return "QUEST NOT COMPLETED";
+        }
+        return tasks.get(currentTaskIndex).getDescription();
+    }
+
+    /**
+     * Returns the hint for the current subtask being completed
+     * for a test or "QUEST COMPLETED" if the quest is completed.
+     */
+    public String getCurrentTaskHint() {
+        if (isQuestCompleted()) {
+            return "QUEST COMPLETED";
+        }
+        if (currentTaskIndex == getNumQuestTasks()) {
+            return "QUEST NOT COMPLETED";
+        }
+        return tasks.get(currentTaskIndex).getHint();
+    }
+    /** Progress (increments) number of quest subtasks completed. */
+    public void progressQuest() {
+        if (!isQuestCompleted() && !isFailed) {
+            if(taskCompletionTriggers!=null){
+                ServiceLocator.getEventService().getGlobalEventHandler().trigger(taskCompletionTriggers[currentTaskIndex]);
+            }
+            currentTaskIndex++;
+        }
+        if(isQuestCompleted()){
+            this.isActive = false;
+            if(taskCompletionTriggers!=null && taskCompletionTriggers.length != 0){
+                    ServiceLocator.getEventService().getGlobalEventHandler().trigger(taskCompletionTriggers[taskCompletionTriggers.length - 1]);
```

Comment:
It would be better for readability to break this command up over two or three lines.

Reply:
Thank you. Fixing now.